nvflare/app_common/abstract/learner2.py

```diff
@@ -0,0 +1,149 @@
+
+
+class Learner2(FLComponent):
+
+    STATE = None
+
+    def __init__(self):
+        super().__init__()
+        self.engine = None
+        self.fl_ctx = None
+        self.workspace = None
+        self.shareable = None
+        self.args = None
+        self.site_name = None
+        self.job_id = None
+        self.app_root = None
+        self.abort_signal = None
+        self.current_round = 0
+        self.total_rounds = 0
+
+    def is_aborted(self) -> bool:
+        return self.abort_signal and self.abort_signal.triggered
+
+    def get_shareable_header(self, key: str, default=None):
+        if not self.shareable:
+            return default
+        return self.shareable.get_header(key, default)
+
+    def get_context_prop(self, key: str, default=None):
+        if not self.fl_ctx:
+            return default
+        assert isinstance(self.fl_ctx, FLContext)
+        return self.fl_ctx.get_prop(key, default)
+
+    def debug(self, msg: str):
+        self.log_debug(self.fl_ctx, msg)
+
+    def info(self, msg: str):
+        self.log_info(self.fl_ctx, msg)
+
+    def error(self, msg: str):
+        self.log_error(self.fl_ctx, msg)
+
+    def warning(self, msg: str):
+        self.log_warning(self.fl_ctx, msg)
+
+    def exception(self, msg: str):
+        self.log_exception(self.fl_ctx, msg)
+
+    def critical(self, msg: str):
+        self.log_critical(self.fl_ctx, msg)
+
+    def panic(self, reason: str):
+        self.system_panic(reason, self.fl_ctx)
+
+    def initialize(self, parts: dict):
+        """Initialize the Learner object. This is called before the Learner can train or validate.
+
+        This is called only once.
+
+        Args:
+            parts: components to be used by the Trainer
+        """
+        pass
+
+    @abstractmethod
+    def train(self, dxo: DXO) -> Union[str, DXO]:
+        """Called to perform training. Can be called many times during the lifetime of the Learner.
+
+        Args:
+            dxo: the training input data (e.g. model weights)
+
+        Returns: train result as a DXO if successful; or return code as str if failed.
+
+        """
+        pass
+
+    def get_model_for_validation(self, model_name: str) -> Union[str, DXO]:
+        """Called to return the trained model from the Learner.
+
+        Args:
+            model_name: type of the model for validation
+
+        Returns: trained model for validation
+
+        """
+        pass
+
+    def validate_before_train(self, dxo: DXO) -> Union[str, DXO]:
+        """Validate the current model with the specified weights in dxo before training.
+
+        Args:
+            dxo: the DXO object that contains model weights
+
+        Returns: validation metrics in DXO if successful; or return code if failed.
+
+        """
+        pass
+
+    def validate(self, dxo: DXO, validate_type: str, model_owner: str) -> Union[str, DXO]:
+        """Validate the model with the specified weights in dxo
```

Comment:
Could we set the "validate_type" and "model_owner" in dxo's metadata?
Then this signature could be `def validate(self, dxo)`

Reply:
If we want to hide it, it will go to "self". We shouldn't change the DXO that comes from the server.
Besides, validate_type and model_owner is not a FLARE specific concept, and it is essential to the validate operation, I think we should keep them here.

Reply:
I guess it comes down to a more general question, what do we think should be in DXO?

Should the "task" specific stuffs (like here "validate_type" and "model_owner") be in the DXO?
How does DXO different from Shareable?

Reply:
DXO should be something that only describes the data. One rule of thumb to use: can the data to saved to disk and consumed by a different app independently? In this regard, "model_owner" may qualify to be included into the meta of the DXO since you can say that the owner of the data  should go with the data; whereas "validate_type" is just an instruction to the client how to do the validate, which is not part of the data.